core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala

```diff
@@ -0,0 +1,378 @@
+   */
+  def askWithReply[T: ClassTag](message: Any): T
+
+  /**
+   * Send a message to the corresponding [[RpcEndpoint.receive]] and get its result within a
+   * specified timeout, throw a SparkException if this fails even after the specified number of
+   * retries.
+   *
+   * Note: this is a blocking action which may cost a lot of time, so don't call it in an message
+   * loop of [[RpcEndpoint]].
+   *
+   * @param message the message to send
+   * @param timeout the timeout duration
+   * @tparam T type of the reply message
+   * @return the reply message from the corresponding [[RpcEndpoint]]
+   */
+  def askWithReply[T: ClassTag](message: Any, timeout: FiniteDuration): T
+
+  /**
+   * Sends a one-way asynchronous message. Fire-and-forget semantics.
+   */
+  def send(message: Any): Unit
+
+  /**
+   * Send a message to the corresponding [[RpcEndpoint.receiveAndReply]] asynchronously.
+   * Fire-and-forget semantics.
+   *
+   * The receiver will reply to sender's [[RpcEndpoint.receive]] or [[RpcEndpoint.receiveAndReply]]
+   * depending on which one of [[RpcResponse.reply]]s is called.
+   */
+  def sendWithReply(message: Any, sender: RpcEndpointRef): Unit
+
+  /**
+   * Send a message to the corresponding [[RpcEndpoint.receiveAndReply)]] and return a `Future` to
+   * receive the reply within a default timeout.
+   */
+  def sendWithReply[T: ClassTag](message: Any): Future[T]
+
+  /**
+   * Send a message to the corresponding [[RpcEndpoint.receiveAndReply)]] and return a `Future` to
+   * receive the reply within the specified timeout.
+   */
+  def sendWithReply[T: ClassTag](message: Any, timeout: FiniteDuration): Future[T]
+
+  def toURI: URI
+}
+
+/**
+ * Represent a host with a port
+ */
+private[spark] case class RpcAddress(host: String, port: Int) {
+  // TODO do we need to add the type of RpcEnv in the address?
+
+  val hostPort: String = host + ":" + port
+
+  override val toString: String = hostPort
+}
+
+private[spark] object RpcAddress {
+
+  /**
+   * Return the [[RpcAddress]] represented by `uri`.
+   */
+  def fromURIString(uri: String): RpcAddress = {
+    val u = new java.net.URI(uri)
+    RpcAddress(u.getHost, u.getPort)
+  }
+
+  def fromSparkURL(sparkUrl: String): RpcAddress = {
+    val (host, port) = Utils.extractHostPortFromSparkUrl(sparkUrl)
+    RpcAddress(host, port)
+  }
+}
+
+/**
+ * Indicate that a new connection is established.
+ *
+ * @param address the remote address of the connection
+ */
+private[spark] case class AssociatedEvent(address: RpcAddress)
+
+/**
+ * Indicate a disconnection from a remote address.
+ *
+ * @param address the remote address of the connection
+ */
+private[spark] case class DisassociatedEvent(address: RpcAddress)
+
+/**
+ * Indicate a network error.
+ * @param address the remote address of the connection which this error happens on.
+ * @param cause the cause of the network error.
+ */
+private[spark] case class NetworkErrorEvent(address: RpcAddress, cause: Throwable)
```

Comment:
Do you mean you like adding the methods in the previous NetworkRpcEndpoint to RpcEndpoint?